solutions/silver/baltic-12-mobile.mdx

```diff
@@ -100,3 +98,136 @@ int main() {
+
+![Maximum Point](<baltic-12-mobile/perpendicular.png>)
+
+This intersection can be represented by $\frac{b_x^2 + b_y^2 - a_x^2 - a_y^2}{2(b_x - a_x)}$.
+
+From here on, we'll refer to this point as $p(a, b)$, where $a$ and $b$ are stations.
+
+</Spoiler>
+
+Now, let's consider cases where there exist unnecessary stations.
+
+Without loss of generality, let's assume all points are above the y-axis, and $a_x < b_x < c_x$.
+
+### Case 1: $p(a, b) < p(b, c)$
+
+In this arrangement, all stations are the closest to some points on the highway, so we should keep them.
+
+![Case 1](<baltic-12-mobile/case1.png>)
+
+### Case 2: $p(a, b) > p(b, c)$
+
+Note how this creates a cross, which makes $b$ useless. All points on the highway are now closer to $a$ or $c$. To find the maximum point here, we can just remove $b$, and recalculate for $p(a, c)$.
+
+![Case 2](<baltic-12-mobile/case2.png>)
+
+Since we only remove the most recent stations, this motivates the use of a [stack](/gold/stacks) to perform these removals/additions in $\mathcal{O}(1)$.
+
+## Implementation
+
+**Time Complexity**: $\mathcal{O}(N)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+#define sz(x) (int)x.size()
+
+struct Point {
+	double x, y;
+};
+
+/**
+ * @return the intersection of the perpendicular line that
+ * crosses the midpoint of Points a & b with the highway
+ */
+double max_point(const Point& a, const Point& b) {
+	return (b.x * b.x + b.y * b.y - a.x * a.x - a.y * a.y) /
+	       (2 * b.x - 2 * a.x);
+}
+/** @return the euclidean distance between points a & b */
+double dist(const Point& a, const Point& b) {
+	return sqrt((a.x - b.x) * (a.x - b.x) + (a.y - b.y) * (a.y - b.y));
+}
+
+int main() {
+	int n, l;
+	cin >> n >> l;
+
+	deque<Point> stck;
+	for (int i = 0; i < n; i++) {
+		Point cur;
+		cin >> cur.x >> cur.y;
+		cur.y = abs(cur.y);
+
+		// always take the one with the lowest y coord
+		if (sz(stck) && stck.back().x == cur.x) {
+			if (cur.y >= stck.back().y) {
+				continue;
+			} else if (cur.y < stck.back().y) {
+				stck.pop_back();
+			}
+		}
+
+		// find "crosses"
+		while (sz(stck) > 1 && max_point(stck[sz(stck) - 2], stck.back()) >
+		                           max_point(stck.back(), cur)) {
+			stck.pop_back();
+		}
+
+		stck.push_back(cur);
+	}
+
+	// out of bounds
+	while (sz(stck) > 1 && max_point(stck[0], stck[1]) < 0) stck.pop_front();
+	while (sz(stck) > 1 && max_point(stck[sz(stck) - 2], stck.back()) > l)
+		stck.pop_back();
+
+	double ans = 0;
+	for (int x = 0; x < sz(stck); x++) {
+		Point left = {(x ? max_point(stck[x], stck[x - 1]) : 0), 0};
+		Point right = {(x < sz(stck) - 1 ? max_point(stck[x], stck[x + 1]) : l),
+		               0};
+
+		if (left.x < 0 || right.x > l || right.x < 0 || left.x > l) continue;
```

Comment:
maybe some comments here
also since the formatting for `right` is really ugly, try breaking the construction up into variables

Reply:
oops ty, totally forgot